skfda/misc/operators/_linear_differential_operator.py

```diff
@@ -585,42 +635,76 @@ def fdatagrid_penalty_matrix_optimized(
     """
     Optimized version for FDatagrid.
 
-    If the data_matrix of the FDataGrid is the identity, the resulting
-    matrix will be the gram matrix for functions discretized in the
-    given grid points. That is to say, in order to calculate the norm
+    The datamatrix of the given FDataGrid has to be the identity matrix,
+    the resulting matrix will be the gram matrix for functions discretized
+    in the given grid points. That is to say, in order to calculate the norm
     of the linear operator applied to functions discretized in the given
     grid points, it is enough to multiply the returned matrix by the
     values in the points of the grid on both sides.
 
-    If the data_matrix of the FDataGrid is not the identity, the resulting
-    matrix will be the gram matrix for functions expressed as a combination
-    of the given set of functions (the entries of the fdatagrid). The
-    intended use is to calculate the norm of the linear operator for
-    functions expressed in a basis that is a FDataGrid (see CustomBasis).
-
-    Note that the first case is a particular case of the second one.
     """
-    evaluated_basis = sum(
-        w(basis.grid_points[0]) if callable(w) else w
-        * basis.derivative(order=i)(basis.grid_points[0])
-        for i, w in enumerate(linear_operator.weights)
-    )
-    assert not isinstance(evaluated_basis, int)
-
-    indices = np.triu_indices(basis.n_samples)
-    product = evaluated_basis[indices[0]] * evaluated_basis[indices[1]]
+    max_deriv = len(linear_operator.weights) - 1
+    n_points = basis.grid_points[0].shape[0]
+    findiff_accuracy = 2
 
-    triang_vec = scipy.integrate.simps(product[..., 0], x=basis.grid_points)
-
-    matrix = np.empty((basis.n_samples, basis.n_samples))
-
-    # Set upper matrix
-    matrix[indices] = triang_vec
+    evaluated_basis = _optimized_operator_evaluation_in_grid(
+        linear_operator,
+        basis.grid_points[0],
+        findiff_accuracy=findiff_accuracy,
+    )
 
-    # Set lower matrix
-    matrix[(indices[1], indices[0])] = triang_vec
+    # The support of the result of applying the linear operator
+    # is a small subset of the grid points. We calculate the
+    # maximum number of points to the left and right of the
+    # point where the linear operator is applied that can be
+    # different from zero.
+
+    # Formula derived from the definition of coefficients_not_uni
+    # in findiff.
+    spread = 2 * math.floor((max_deriv + 1) / 2) - 1
+    # The accuracy parameter in findiff increases the number of
+    # points considered to the left and the right
+    spread += findiff_accuracy + 1
```

Comment:
I don't understand the question. Could you expand on it?

Reply:
I think that the spread can be computed from the offsets already computed, without repeating the internal logic of `findiff` here.

Reply:
Oh, I didn't think of that. I think it should be possible and it would make it safer in terms of making sure that this code is still applicable even if that internal logic changes. I'll change it and test it.

Reply:
Unfortunately, that is not possible. It took me a bit to remember why, but the issue is that the matrix is constructed point by point. That is to say, if you look at the final matrix returned by the function (it is transposed at the end), the offsets will tell you how many elements are not 0 column-wise, not row-wise.

Reply:
Nevermind. It is possible. It's not entirely straightforward, and I couldn't find a way to do it initially. I now have a solution for this and I have implemented it in 6d60b05417cd85aec81439a78711f555cc975272.